command/connect/envoy/exec_unix.go

```diff
@@ -0,0 +1,123 @@
+// +build linux darwin
+
+package envoy
+
+import (
+	"errors"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"golang.org/x/sys/unix"
+)
+
+func execEnvoy(binary string, prefixArgs, suffixArgs []string, bootstrapJson []byte) error {
+	// Write the Envoy bootstrap config file out to disk in a pocket universe
+	// visible only to the current process (and exec'd future selves).
+	fd, err := writeEphemeralEnvoyTempFile(bootstrapJson)
+	if err != nil {
+		return errors.New("Could not write envoy bootstrap config to a temp file: " + err.Error())
+	}
+
+	// On unix systems after exec the file descriptors that we should see:
+	//
+	//   0: stdin
+	//   1: stdout
+	//   2: stderr
+	//   ... any open file descriptors from the parent without CLOEXEC set
+	//
+	// Above we explicitly disabled CLOEXEC for our temp file, so assuming
+	// FD numbers survive across execs, it should just be the value of
+	// `fd`. This is accessible as a file itself (trippy!) under
+	// /dev/fd/$FDNUMBER.
+	magicPath := filepath.Join("/dev/fd", strconv.Itoa(int(fd)))
+
+	// First argument needs to be the executable name.
+	envoyArgs := []string{binary}
+	envoyArgs = append(envoyArgs, prefixArgs...)
+	envoyArgs = append(envoyArgs, "--v2-config-only",
+		"--disable-hot-restart",
+		"--config-path",
+		magicPath,
+	)
+	envoyArgs = append(envoyArgs, suffixArgs...)
+
+	// Exec
+	if err = unix.Exec(binary, envoyArgs, os.Environ()); err != nil {
+		return errors.New("Failed to exec envoy: " + err.Error())
+	}
+
+	return nil
+}
+
+func writeEphemeralEnvoyTempFile(b []byte) (uintptr, error) {
+	f, err := ioutil.TempFile("", "envoy-ephemeral-config")
```

Comment:
`TempFile` does currently create files with 0600 which is good but it doesn't explicitly guarantee that. It would be wild if they ever made it more permissive but is it worth being defensive and doing it long-hand just to guarantee it gets the permissions we want? Totally reasonable to leave it for now just a thought.